### PR TITLE
Grid search fix

### PR DESCRIPTION
**Pull request by mrghg**
Grid search updated to allow arbitrary keyword arguments to train_mlp to be iterated over (e.g., ```time_shift_hours```, ```balance```). 

You need to specify if any of these ```data_kwargs``` also change the validation set via the ```validation_keys``` argument.

(also added some half-finished functionality to the features plotting)

### CODE REVIEW

File: ml_baselines/modelling/train.py
Comment (elenafillo) [inline ml_baselines/modelling/train.py:219]: need to add some default parameters to use when mlp_params=None

**Review by elenafillo — CHANGES_REQUESTED** [review]
looks good! Main things to change are adding default parameters to train_mlp and correcting the format of the hidden layer sizes. after that, it's ready to merge!

File: ml_baselines/modelling/train.py [hunk 1]
```diff
@@ -266,57 +266,82 @@ def train_mlp(site,
...
+            'early_stopping': [True],
             'shuffle': [False]
-            # 'activation': ['relu'],
-            # 'solver': ['adam'],
-            # 'alpha': [0.0001, 0.05],
-            # 'learning_rate': ['constant','adaptive'],
-            # 'batch_size': [100, 200, 300],
-            # 'max_iter': [1000, 2000],
-            # 'early_stopping': [True, False]
         }
```
[the remark this answers is not in the document]
Comment (mrghg) [reply ml_baselines/modelling/train.py:316]:
```suggestion
    if param_grid is None:
        param_grid = {
            'hidden_layer_sizes': [(50),],
            'activation': ['relu'],
            'solver': ['adam'],
            'alpha': [0.0001],
            'batch_size': [5, 10],
            'max_iter': [1000],
            'early_stopping': [True],
            'shuffle': [False]
        }
```

[the remark this answers is not in the document]
Comment (mrghg) [reply ml_baselines/modelling/train.py:316]: OK, changed

File: ml_baselines/modelling/train.py [hunk 2]
```diff
@@ -266,57 +266,82 @@ def train_mlp(site,
...
+        X_train, y_train = get_train_test_data(site, "train", **combo_kw)
+        X_val, y_val = get_train_test_data(site, "validation", **val_kw)
+
+        # Combine training and validation sets; use PredefinedSplit so validation
+        # rows are never used for fitting during cross-validation
         X_all = pd.concat([X_train, X_val])
         y_all = pd.concat([y_train, y_val])
-
-        # Create a test_fold array: assign -1 for training rows and 0 for validation rows
         test_fold = [-1] * len(X_train) + [0] * len(X_val)
```
[the remark this answers is not in the document]
Comment (mrghg) [reply ml_baselines/modelling/train.py:347]:
```suggestion
    for combo in combos:
        combo_kw = dict(zip(keys, combo))
        print(f"... data kwargs: {combo_kw}")

        val_kw = {k: combo_kw[k] for k in validation_keys if k in combo_kw}
        X_train, y_train = get_train_test_data(site, "train", **combo_kw)
        X_val, y_val = get_train_test_data(site, "validation", **val_kw)

        # Combine training and validation sets; use PredefinedSplit so validation
        # rows are never used for fitting during cross-validation
        X_all = pd.concat([X_train, X_val])
        y_all = pd.concat([y_train, y_val])
        test_fold = [-1] * len(X_train) + [0] * len(X_val)
```

[the remark this answers is not in the document]
Comment (mrghg) [reply ml_baselines/modelling/train.py:347]: OK, added

File: ml_baselines/modelling/train.py [hunk 3]
```diff
@@ -336,39 +361,36 @@ def train_mlp_grid_search(site,
...
+    f1_val = f1_score(y_val_final, pred_val)
 
     print(f"Validation Precision = {precision_val:.3f}")
     print(f"Validation Recall = {recall_val:.3f}")
     print(f"Validation F1 Score = {f1_val:.3f}")
 
-    # # Save model
-    # met_str = f"-{cfg.met_type}" if cfg.met_type else ""
-    # joblib.dump(best_model, Path(cfg.models_path) / f"best_mlp{met_str}_{site}.joblib")
+    return best_model, best_best_params, best_combo_kw
```
[the remark this answers is not in the document]
Comment (mrghg) [reply ml_baselines/modelling/train.py:397]: Ignore for now. Let's fix this when we make the more general model

File: ml_baselines/modelling/train.py
Comment (mrghg) [reply ml_baselines/modelling/train.py:219]: Agreed. This has been updated in your PR, so should come with the next merge

File: notebooks/plot_features.ipynb [hunk 4]
```diff
@@ -23,105 +23,227 @@
...
+    "    plt.grid()\n",
+    "    plt.show()\n",
+    "\n",
+    "# Interactive widget to select variable\n",
+    "variables = set([col.split('_')[0] for col in df_features.columns])\n",
+    "npoints = len([col for col in df_features.columns if col.startswith(list(variables)[0])])\n",
     "\n",
-    "plot_wind_vectors(time_idx=1,\n",
-    "                  wind_variables=[\"u10\", \"v10\"])"
+    "interact(plot_timeseries, variable=variables, point=IntSlider(min=0, max=int(npoints)-1, step=1, value=0))"
```
[the remark this answers is not in the document]
Comment (mrghg) [reply notebooks/plot_features.ipynb:246]: Ignoring this one for now. This plotting should be fixed in a future PR
